tasks/ast_tools/src/derives/clone_in.rs

```diff
@@ -60,52 +62,92 @@ fn derive_enum(def: &EnumDef) -> TokenStream {
         }
     };
 
-    impl_clone_in(&ty_ident, def.has_lifetime, &alloc_ident, &body)
+    impl_clone_in(&ty_ident, def.has_lifetime, &alloc_ident, &body, &quote!())
 }
 
 fn derive_struct(def: &StructDef) -> TokenStream {
     let ty_ident = def.ident();
-
-    let (alloc_ident, body) = if def.fields.is_empty() {
-        (format_ident!("_"), quote!(#ty_ident))
+    let needs_override_clone_in_with_semantic_ids = Cell::new(false);
+    let (alloc_ident, clone_in_body, clone_in_with_semantic_ids_function) = if def.fields.is_empty()
+    {
+        (format_ident!("_"), quote!(#ty_ident), quote!())
     } else {
-        let fields = def.fields.iter().map(|field| {
-            let ident = field.ident();
-            match field.markers.derive_attributes.clone_in {
-                CloneInAttribute::Default => quote!(#ident: Default::default()),
-                CloneInAttribute::None => {
-                    quote!(#ident: CloneIn::clone_in(&self.#ident, allocator))
+        let alloc_ident = format_ident!("allocator");
+        let clone_in_fields = def
+            .fields
+            .iter()
+            .map(|field| {
+                let ident = field.ident();
+                match field.markers.derive_attributes.clone_in {
+                    CloneInAttribute::Default => {
+                        needs_override_clone_in_with_semantic_ids.set(true);
+                        quote!(#ident: Default::default())
+                    }
+                    CloneInAttribute::None => {
+                        quote!(#ident: CloneIn::clone_in(&self.#ident, allocator))
+                    }
                 }
-            }
-        });
-        (format_ident!("allocator"), quote!(#ty_ident { #(#fields),* }))
+            })
+            .collect::<Vec<_>>()
+            .into_iter();
+        let clone_in_with_semantic_ids_token_stream = if needs_override_clone_in_with_semantic_ids
+            .get()
+        {
+            let fields = def.fields.iter().map(|field| {
+                let ident = field.ident();
+                quote!(#ident: CloneIn::clone_in(&self.#ident, allocator))
```

Comment:
This isn't correct. It should be calling `clone_in_with_semantic_ids`, not `clone_in`.